### PR TITLE
Background Session Tracking

### LINKED ISSUE #449: Improve Session Tracking Reliability
Use a background task to allow the 10 second timer to continue when the app is put in the background, giving the session time to complete.

### PR DESCRIPTION
**Pull request by seanrucker**
Closes #449

### CODE REVIEW

File: Sources/Services/SessionController/SessionController.swift [hunk 1]
```diff
@@ -11,32 +11,37 @@ import UIKit
...
     }
     
     deinit {
-        if let observer = didBecomeActiveObserver {
-            NotificationCenter.default.removeObserver(observer)
-        }
-        if let observer = willResignActiveObserver {
-            NotificationCenter.default.removeObserver(observer)
-        }
+        observers.forEach(NotificationCenter.default.removeObserver)
```
Comment (orospakr) [inline Sources/Services/SessionController/SessionController.swift:44]: clear observers array afterwards?

File: Sources/Services/SessionController/Session.swift [hunk 2]
```diff
@@ -72,21 +72,31 @@ class Session {
         case .started(let startedAt):
             // Capture endedAt now, as opposed to when the timer fires
             let endedAt = Date()
+            
+            var backgroundTaskID: UIBackgroundTaskIdentifier!
+            backgroundTaskID = UIApplication.shared.beginBackgroundTask(withName: "Keep Alive Timer") { [weak self] in
```
Comment (orospakr) [inline Sources/Services/SessionController/Session.swift:78]: I think a comment explaining what you just explained to me would be helpful: that we want a chance to expire our sessions without waiting for the app to be brought back into the foreground sometime in an arbitrarily distant future. If I have that right.

**Review by orospakr — CHANGES_REQUESTED** [review]

File: Sources/Services/SessionController/SessionController.swift (hunk 1 above)
Comment (seanrucker) [reply Sources/Services/SessionController/SessionController.swift:44]: Good catch.

File: Sources/Services/SessionController/Session.swift (hunk 2 above)
Comment (seanrucker) [reply Sources/Services/SessionController/Session.swift:78]: You do, however I was explaining to you what the Apple docs arguably do a better job of. I.e. if you understand how Apple's background tasks work, then this code is very straightforward.

Comment (orospakr) [reply Sources/Services/SessionController/Session.swift:78]: I'm thinking more of how it interacts with our expiry algorithm.  eg.:

    // Best-effort attempt to expire our background sessions before app is killed in the background.

**Review by orospakr — APPROVED** [review]

**Pull request merged**